docs/specs/parsing.md

```diff
@@ -1,109 +1,242 @@
-  parsing needs to consume
-
-## Focus on UTF8
-
-While we generally need to handle parsing for more than just UTF8 we belive it's
-beneficial to have an implementation that is hand-optimized for UTF8 instead of
-having an implementation that is going through some abstraction.
+The current .NET Framework parsing APIs (e.g. int.TryParse) can parse System.String (UTF16) text.
+```c#
+string text = ...
+int value;
+if(int.TryParse(text, out value)){
+   ...
+}
+```
+These APIs work great in many scenarios, e.g. parsing text contained in GUI application text boxes. But they are completely not suitable for processing modern network protocols, which are often text (e.g. JSON, HTTP headers), but contained in byte buffers, not strings, and encoded with UTF8/ASCII, not UTF16. Because of this, all modern web servers written for the .NET platform either don't use the current BCL parsing APIs or take performance hit of transcoding from UTF8 to UTF16, and copying from buffers to strings.
+```c#
+byte[] json = ...
+string text = Encoding.UTF8.GetString(json); // this is very expensive: copy, transformation, object allocations
+int value;
+if(int.TryParse(text, out value)){
+   ...
+}
+```
+To address modern networking scenarios beter, we will provide parsing APIs that:
+a) Can parse byte buffers, with out the need to have a string to parse
+b) Can parse text encoded as UTF8 (and possibly other encodings)
+c) Can parse without any GC heap allocations
+
+This document spells out the details of the project.
+
+## API Design
+We will provide low level APIs that can parse byte\* and ReadOnlySpan\<byte\> buffers (moral equivalent of int.TryParse), and higher level APIs to parse sequences of such buffers (similar to BinaryReader for encoded text streams). 
+
+To familiarize yourself with ReadOnlySpan\<T\> refer to: https://github.com/dotnet/corefxlab/blob/master/docs/specs/span.md
+
+### Low Level APIs
+
+The low level APIs are policy free (not specific to scenarios, higher level frameworks, etc.) and optimized for speed, and will look like the following (per data type being parsed; the sample shows Int32 APIs):
+```c#
+public static class PrimitiveParser {
+    // most general "overload"
+    public static bool TryParseInt32(ReadOnlySpan<byte> text, out int value, out int bytesConsumed, EncodingData encoding=null, TextFormat format=null);
+    
+    // UTF8-invariant overloads
+    public static class InvariantUtf8 {
+       
+        // We did some preliminary measurements, and the shorter overload is non-trivially faster. We should double check though.
+        public unsafe static bool TryParseInt32(byte* text, int length, out int value);
+        public unsafe static bool TryParseInt32(byte* text, int length, out int value, out int bytesConsumed);
+        
+        // overloads optimized for hexadecimal numbers
+        public static class Hex {
+            public static bool TryParseInt32(ReadOnlySpan<byte> text, out int value);
```

Comment:
There is a mix of ReadOnlySpan and unsafe pointers. Is it intentional? I do not see any pattern in it.